File: backend/src/main/java/mk/uikm/finki/emtlab/model/domain/Book.java
package mk.uikm.finki.emtlab.model.domain;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import mk.uikm.finki.emtlab.model.enums.BookCategory;
import mk.uikm.finki.emtlab.model.enums.BookState;
import java.time.LocalDate;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "books")
@NamedEntityGraph(
        name = "book-entity-graph",
        attributeNodes = {
                @NamedAttributeNode(value = "author", subgraph = "author-graph")
        },
        subgraphs = {
                @NamedSubgraph(
                        name = "author-graph",
                        attributeNodes = {
                                @NamedAttributeNode("country")
                        }
                )
        }
)
public class Book extends BaseAuditableEntity{
    @Column(nullable = false)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BookCategory category;

    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    private Author author;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BookState state;

    @Column(name = "available_copies", nullable = false)
    private Integer availableCopies;

    @Column(nullable = false)
    private boolean deleted = false;

    @Column(name = "date_published", nullable = false)
    private LocalDate datePublished;

    public boolean getDeleted() {
        return this.deleted;
    }
}

File: frontend/src/api/types/book.ts
import type { Author } from './author';

export const BookCategory = {
    NOVEL: 'NOVEL',
    THRILLER: 'THRILLER',
    HISTORY: 'HISTORY',
    FANTASY: 'FANTASY',
    BIOGRAPHY: 'BIOGRAPHY',
    CLASSICS: 'CLASSICS',
    DRAMA: 'DRAMA',
} as const;
export type BookCategory = (typeof BookCategory)[keyof typeof BookCategory];

export const BookState = {
    GOOD: 'GOOD',
    BAD: 'BAD',
} as const;
export type BookState = (typeof BookState)[keyof typeof BookState];


export interface Book {
    id: number,
    name: string;
    category: BookCategory;
    authorId: number;
    state: BookState;
    availableCopies: number;
    datePublished: Date;
    deleted: false;
}

export interface BookDetails {
    id: number,
    name: string;
    category: BookCategory;
    author: Author;
    state: BookState;
    availableCopies: number;
    datePublished: Date;
}

export interface BookFormData {
    name: string;
    category: BookCategory;
    authorId: number;
    state: BookState;
    availableCopies: number;
    datePublished: Date;
    deleted: boolean;
}
